### PR TITLE
gadget, snap/pack: perform extended validation of gadget metadata and contents

### PR DESCRIPTION
**Pull request by bboozzoo**
Extend the validation of gadget snaps to cover gadget metadata and the contents of a snap. Error in metadata or missing content may cause the snap fail during gadget update. Try to catch problems early.

### CODE REVIEW

File: gadget/validate.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	if err != nil {
+		return fmt.Errorf("cannot validate gadget metadata: %v", err)
+	}
+
+	for name, vol := range info.Volumes {
+		lv, err := LayoutVolume(gadgetSnapRootDir, &vol, defaultConstraints)
+		if err != nil {
+			return fmt.Errorf("cannot validate volume %q layout: %v", name, err)
+		}
+		if err := validateVolumeContentsPresence(gadgetSnapRootDir, lv); err != nil {
```
Comment (mvo5) [inline gadget/validate.go]: Nice!

File: gadget/validate.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+			continue
+		}
+		for _, c := range s.Content {
+			realSource := filepath.Join(gadgetSnapRootDir, c.Source)
+			if !osutil.FileExists(realSource) {
+				return fmt.Errorf("cannot validate structure %v, content %v: source path does not exist", s, c)
+			}
+			if strings.HasSuffix(c.Source, "/") {
+				// expecting a directory
+				if err := checkSourceIsDir(realSource + "/"); err != nil {
```
Comment (stolowski) [inline gadget/validate.go]: Is the `+ "/"` neccessary?

Comment (bboozzoo) [reply gadget/validate.go]: Yeah, preserving the input form (which was cleaned up by filepath.Join()) helps to produce a nicer/more meaningful error message.

File: gadget/validate.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	// bare structure content is checked to exist during layout
+	// make sure that filesystem content source paths exist as well
+	for _, s := range vol.LaidOutStructure {
+		if s.IsBare() {
+			continue
+		}
+		for _, c := range s.Content {
+			realSource := filepath.Join(gadgetSnapRootDir, c.Source)
+			if !osutil.FileExists(realSource) {
+				return fmt.Errorf("cannot validate structure %v, content %v: source path does not exist", s, c)
```
Comment (pedronis) [inline gadget/validate.go]: if I read the code correctly:  we'll get "cannot validate... cannot validate" error messages... in general we don't use "cannot validate" form of message, more like "invalid ..."

File: gadget/validate.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+		return fmt.Errorf("cannot validate gadget metadata: %v", err)
+	}
+
+	for name, vol := range info.Volumes {
+		lv, err := LayoutVolume(gadgetSnapRootDir, &vol, defaultConstraints)
+		if err != nil {
+			return fmt.Errorf("cannot validate volume %q layout: %v", name, err)
+		}
+		if err := validateVolumeContentsPresence(gadgetSnapRootDir, lv); err != nil {
+			return fmt.Errorf("cannot validate volume %q: %v", name, err)
```
Comment (pedronis) [inline gadget/validate.go]: see my comment about being careful about getting errors that are a recursive chain of "cannot validate..."

File: gadget/validate_test.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        filesystem: ext4
+        content:
+          - source: foo/
+            target: /
+
+`
+	makeSizedFile(c, filepath.Join(s.dir, "meta/gadget.yaml"), 0, []byte(gadgetYamlContent))
+
+	err := gadget.Validate(s.dir)
+	c.Assert(err, ErrorMatches, `cannot validate volume "bad": cannot validate structure #0 \("bad-struct"\), content source:foo/: source path does not exist`)
```
Comment (pedronis) [inline gadget/validate_test.go]: as I said:

"cannot validate ..: invalid structure"

or even

"invalid volume "bad": invalid structure..."

would be less clunky (I think)

File: overlord/devicestate/firstboot_test.go [hunk 6]
```diff
@@ -1160,14 +1160,12 @@ type: kernel`
 		kernelFname = fname
 	}
 
-	if !opts.classic {
```
Comment (pedronis) [inline overlord/devicestate/firstboot_test.go]: this change looks strange, why do we need volumes for classic now?

File: snap/pack/pack.go [hunk 7]
```diff
@@ -116,6 +117,12 @@ func loadAndValidate(sourceDir string) (*snap.Info, error) {
 	if err := snap.ValidateContainer(snapdir.New(sourceDir), info, logger.Noticef); err != nil {
 		return nil, err
 	}
+
+	if info.SnapType == snap.TypeGadget {
+		if err := gadget.Validate(sourceDir, &gadget.ModelConstraints{Classic: true}); err != nil {
```
Comment (pedronis) [inline snap/pack/pack.go]: why Classic: true instead of nil?

File: seed/seed16_test.go [hunk 8]
```diff
@@ -314,6 +314,23 @@ version: 1.0
...
+`
+
+var pcGadgetFiles = [][]string{
+	{"meta/gadget.yaml", pcGadgetYaml},
+}
+
+var snapFiles = map[string][][]string{
+	"pc":               pcGadgetFiles,
+	"pc=18":            pcGadgetFiles,
+	"classic-gadget":   pcGadgetFiles,
```
Comment (pedronis) [inline seed/seed16_test.go]: why do we need this now?

File: overlord/devicestate/firstboot_test.go (hunk 6 above)
Comment (bboozzoo) [reply overlord/devicestate/firstboot_test.go]: Reverted along with the classic-gadget in seed tests. This was left behind from the initial version of the code.

File: snap/pack/pack.go (hunk 7 above)
Comment (bboozzoo) [reply snap/pack/pack.go]: Switched to nil. Wonder whether we should have a way for figuring out the model constraints somehow at the time when you pack the snap.

File: image/image_test.go [hunk 9]
```diff
@@ -452,20 +452,38 @@ func (s *imageSuite) TestHappyDecodeModelAssertion(c *C) {
...
+			{"grub.conf", ""}, {"grub.cfg", "I'm a grub.cfg"},
+			{"meta/gadget.yaml", pcGadgetYaml},
+		}, snap.R(4), publishers["pc18"])
 	}
 
 	if _, ok := publishers["classic-gadget"]; ok {
-		s.MakeAssertedSnap(c, packageClassicGadget, [][]string{{"some-file", "Some file"}}, snap.R(5), publishers["classic-gadget"])
+		s.MakeAssertedSnap(c, packageClassicGadget, [][]string{
+			{"some-file", "Some file"},
+			{"meta/gadget.yaml", pcGadgetYaml},
```
Comment (pedronis) [inline image/image_test.go]: we don't need this change I suppose

File: image/image_test.go [hunk 10]
```diff
@@ -452,20 +452,38 @@ func (s *imageSuite) TestHappyDecodeModelAssertion(c *C) {
...
+		}, snap.R(5), publishers["classic-gadget"])
 	}
 
 	if _, ok := publishers["classic-gadget18"]; ok {
-		s.MakeAssertedSnap(c, packageClassicGadget18, [][]string{{"some-file", "Some file"}}, snap.R(5), publishers["classic-gadget18"])
+		s.MakeAssertedSnap(c, packageClassicGadget18, [][]string{
+			{"some-file", "Some file"},
+			{"meta/gadget.yaml", pcGadgetYaml},
```
Comment (pedronis) [inline image/image_test.go]: same

**Pull request merged**